scripts/rest_service.py

```diff
@@ -0,0 +1,300 @@
+
+        name: name of backend psij executor to use for this client
+              defaults to `local`
+        url : url to pass to backend psij executor
+              defaults to `fork://localhost/`
+
+        returns: a unique ID to identify the client on further requests.
+
+        Register client, configure the service's psij executor to use for this
+        client, and return a unique client ID.  That ID is required for
+        all further requests.
+
+
+    submit(cid: str, spec: Dict[str, Any]) -> str
+
+        cid : client ID obtained via `register_client`
+        spec: serialized psij.JobSepc
+
+        This method submits a job as described by the JobSpec to the backend psij
+        executor and returns the job ID.
+
+
+    cancel(cid: str, jobid: str) -> None
+
+        cid  : client ID obtained via `register_client`
+        jobid: job ID obtained via `submit` or `list`
+
+        This method will cancel the specified job.  The method returns without
+        waiting for the cancelation request to suceed, the callee should observe
+        state notifications to confirm successfull cancellation.
+
+
+    list(cid: str) -> List[str]
+
+        cid  : client ID obtained via `register_client`
+
+        This method will return a list of job IDs known to this service.
+'''
+
+
+from fastapi import FastAPI, WebSocket, WebSocketDisconnect
+
+import psij
+import json
+import queue
+import asyncio
+import logging
+import functools
+
+from typing import List, Dict, Any, Optional
+
+
+# ------------------------------------------------------------------------------
+#
+class _Client(object):
+
+    def __init__(self, jex: psij.JobExecutor):
+
+        self.jex: psij.JobExecutor = jex
+        self._jobs: Dict[str, psij.Job] = dict()
+        self.ws: Optional[WebSocket] = None
+        self._queue: queue.Queue = queue.Queue()
+
+    def add_job(self, job: psij.Job) -> None:
+
+        self._jobs[job.id] = job
+
+    def get_job(self, jobid: str) -> Optional[psij.Job]:
+
+        return self._jobs.get(jobid)
+
+    def list_jobs(self) -> List[str]:
+
+        return list(self._jobs.keys())
+
+    def get_msg(self) -> Any:
+        try:
+            return self._queue.get_nowait()
+        except Exception:
+            pass
+
+    def send(self, msg: Any) -> None:
+        self._queue.put(msg)
+
+
+# ------------------------------------------------------------------------------
+#
+class Service(object):
+
+    def __init__(self, app: FastAPI) -> None:
+
+        self._clients: Dict[str, _Client] = dict()
+        self._deserialize = psij.Import()
+        self._log = logging.getLogger('psij')
+        self._cnt: int = 0
+
+        # ----------------------------------------------------------------------
+        # websocket endpoint at which client can regoister for state updates
+        @app.websocket("/ws/{cid}")
+        async def ws_endpoint(ws: WebSocket, cid: str) -> None:
```

Comment:
Is this async function used anywhere? I don't see it referenced anywhere. Is the decorator doing some magic?

Reply:
Indeed, the decorator does some magic - the function is handled in the `app` event loop.